sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2018-2020 VMware, Inc. All Rights Reserved.
+ * SPDX-License-Identifier: BSD-2
+ */
+package com.vmware.ddlog;
+
+import ddlogapi.DDlogAPI;
+import ddlogapi.DDlogException;
+import ddlogapi.DDlogRecord;
+import org.jooq.DSLContext;
+import org.jooq.Field;
+import org.jooq.Record;
+import org.jooq.Result;
+import org.jooq.Table;
+import org.jooq.impl.DSL;
+import org.jooq.tools.jdbc.MockDataProvider;
+import org.jooq.tools.jdbc.MockExecuteContext;
+import org.jooq.tools.jdbc.MockResult;
+
+import javax.annotation.Nullable;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DDlogJooqProvider implements MockDataProvider {
+    private static final String INTEGER_TYPE = "java.lang.Integer";
+    private static final String STRING_TYPE = "java.lang.String";
+    private static final String BOOLEAN_TYPE = "java.lang.Boolean";
+    private static final String LONG_TYPE = "java.lang.Long";
+
+    private final DDlogAPI dDlogAPI;
+    private final DSLContext dslContext;
+    private final Map<String, List<Field<?>>> tables = new HashMap<>();
+
+    public DDlogJooqProvider(final DDlogAPI dDlogAPI, final List<String> sqlStatements) {
+        this.dDlogAPI = dDlogAPI;
+        dslContext = DSL.using("jdbc:h2:mem:");
+        for (final String sql : sqlStatements) {
+            dslContext.execute(sql);
+        }
+        for (final Table<?> table: dslContext.meta().getTables()) {
+            if (table.getSchema().getName().equals("PUBLIC")) {
+                tables.put(table.getName(), Arrays.asList(table.fields()));
+            }
+        }
+    }
+
+    @Override
+    public MockResult[] execute(final MockExecuteContext ctx) throws SQLException {
+        final MockResult[] mock = new MockResult[1];
+        // The execute context contains SQL string(s), bind values, and other meta-data
+        final String sql = ctx.sql();
+
+        if (sql.toUpperCase().startsWith("SELECT")) {
+            final String[] s = ctx.sql().toUpperCase().split(" ");
+            if (!s[s.length - 2].equals("FROM")) {
+                throw new SQLException("Statement not supported: " + sql);
+            }
+            final String tableName = s[s.length - 1];
+            final List<Field<?>> fields = tables.get(tableName);
+            final Result<Record> result = dslContext.newResult(fields);
+            try {
+                dDlogAPI.dumpTable("R" + tableName.toLowerCase(), (record, l) -> {
+                    final Record jooqRecord = dslContext.newRecord(fields);
+                    final Object[] returnValue = new Object[fields.size()];
+                    for (int i = 0; i < fields.size(); i++) {
```

Comment:
Assert that the size is the same?

Reply:
What's the correct way to query the size of the ddlog record? I see `record.getStructField(int i)`, but it does not say how many fields the struct has.

Reply:
In a sense, it'd be an internal error (ddlog or our sql-to-ddlog compiler) for record.<size()> to not be equal to fields.size(). And if it is not the case, we'll correctly get an out of bounds exception. Seems excessive to check every record in dumpTable() for their size. Am I missing something?